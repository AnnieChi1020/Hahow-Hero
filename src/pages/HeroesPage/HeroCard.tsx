import React from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import styled from 'styled-components';
import { Hero } from '../../redux/heroSlice';
import Colors from '../../styles/Colors';
import { H2 } from '../../styles/Fonts';

interface HeroCardProps {
    hero: Hero;
}

const Container = styled.div<{ isActive: boolean }>`
    width: 100%;
    padding: 20px;
    border-radius: 12px;
    background-color: ${({ isActive }) =>
        isActive ? Colors.orange : Colors.beige};
    display: grid;
    grid-template-rows: 200px 1fr;
    gap: 20px;
    cursor: pointer;
    transition: transform 200ms ease-in-out;

    :hover {
        transform: scale(1.05);
    }

    h2 {
        color: ${({ isActive }) => (isActive ? Colors.gray2 : Colors.gray6)};
    }
`;

const ImageContainer = styled.div`
    position: relative;
    width: 100%;
    height: 100%;
    border-radius: 12px;
    overflow: hidden;
`;

const Image = styled.img`
    position: relative;
    z-index: 1;
    width: 100%;
    height: 100%;
    object-fit: cover;
`;

const TextContainer = styled.div``;

function HeroCard(props: HeroCardProps) {
    const { hero } = props;
    const { id, name, image } = hero;

    const navigate = useNavigate();
    const { heroId } = useParams();

    const isActive = !!(heroId && heroId === id);

    return (
        <Container
            data-testid="hero-card-container"
            onClick={() => navigate(`/heroes/${id}`)}
            isActive={isActive}
        >
            <ImageContainer>
                <Image src={image} />
            </ImageContainer>
            <TextContainer>
                <H2>{name}</H2>
            </TextContainer>
        </Container>
    );
}

export default HeroCard;
