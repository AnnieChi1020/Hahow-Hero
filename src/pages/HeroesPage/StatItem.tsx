import React from 'react';
import styled from 'styled-components';
import Colors from '../../styles/Colors';
import { H3 } from '../../styles/Typography';
import Button from '../../components/Button';

interface StatItemProps {
    label: string;
    point: number;
    handleIncrement: () => void;
    handleDecrement: () => void;
    restPoint: number;
}

const Container = styled.div`
    width: 100%;
    display: flex;
    align-items: center;
    padding: 12px 0;

    @media (max-width: 768px) {
        justify-content: space-between;
    }
`;

const Text = styled(H3)`
    @media (max-width: 768px) {
        font-size: 18px;
        line-height: 24px;
    }
`;

const StateLabel = styled(Text)`
    width: 120px;
    text-align: left;
`;

const StatePoint = styled(Text)`
    width: 120px;
    text-align: center;
`;

const ButtonText = styled(Text)`
    color: ${Colors.gray1};
`;

function StatItem(props: StatItemProps) {
    const { label, point, handleIncrement, handleDecrement, restPoint } = props;

    return (
        <Container>
            <StateLabel>{label}</StateLabel>
            <Button onClick={handleDecrement} disabled={point <= 0}>
                <ButtonText>-</ButtonText>
            </Button>
            <StatePoint>{point}</StatePoint>
            <Button onClick={handleIncrement} disabled={restPoint <= 0}>
                <ButtonText>+</ButtonText>
            </Button>
        </Container>
    );
}

export default StatItem;
