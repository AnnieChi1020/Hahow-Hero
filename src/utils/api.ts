import { Hero, HeroProfile } from '../redux/heroSlice';

interface HeroProfileRes {
    heroId: string;
    profile: HeroProfile;
}

const HERO_API = process.env.REACT_APP_HERO_API;

export const fetchHeroes = async (): Promise<Hero[]> => {
    const res = await fetch(`${HERO_API}/heroes`);
    if (res.status !== 200) {
        throw new Error('Network response was not ok');
    }
    const data: Hero[] = await res.json();
    return data;
};

export const fetchHeroProfile = async (
    heroId: string
): Promise<HeroProfileRes | undefined> => {
    const res = await fetch(`${HERO_API}/heroes/${heroId}/profile`);

    if (res.status !== 200) {
        throw new Error('Network response was not ok');
    }

    const data = await res.json();
    return data;
};

export const updateHeroProfile = async (
    heroId: string,
    profile: HeroProfile
): Promise<boolean> => {
    let isUpdated = false;

    const res = await fetch(`${HERO_API}/heroes/${heroId}/profile`, {
        method: 'PATCH',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(profile),
    });
    if (res.status === 200) {
        isUpdated = true;
    }

    return isUpdated;
};
