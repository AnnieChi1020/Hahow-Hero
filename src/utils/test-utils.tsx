import React, { PropsWithChildren } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render, RenderOptions } from '@testing-library/react';
import { configureStore, PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import heroReducer, { initialState } from '../redux/heroSlice';
import { RootState, AppStore } from '../redux/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: PreloadedState<RootState>;
    store?: AppStore;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = { hero: initialState },
        // Automatically create a store instance if no store was passed in
        store = configureStore({
            reducer: { hero: heroReducer },
            preloadedState,
        }),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren): JSX.Element {
        return <Provider store={store}>{children}</Provider>;
    }
    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function getWrapper() {
    const preloadedState = { hero: initialState };
    const store: AppStore = configureStore({
        reducer: { hero: heroReducer },
        preloadedState,
    });

    function Wrapper({ children }: PropsWithChildren): JSX.Element {
        return <Provider store={store}>{children}</Provider>;
    }

    return Wrapper;
}
