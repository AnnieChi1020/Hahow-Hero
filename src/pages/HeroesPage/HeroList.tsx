import React from 'react';
import styled from 'styled-components';
import HeroCard from './HeroCard';
import { Hero } from '../../redux/heroSlice';

const HeroList: React.FC<{ heroList: Hero[] }> = (props) => {
    const { heroList } = props;

    return (
        <Container>
            {heroList.map((h) => (
                <HeroCard key={h.id} hero={h} />
            ))}
        </Container>
    );
};

export default HeroList;

const Container = styled.div`
    display: flex;
    gap: 24px;
`;
