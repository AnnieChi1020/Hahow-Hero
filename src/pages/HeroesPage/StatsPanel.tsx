import React from 'react';
import styled from 'styled-components';
import Button from '../../components/Button';
import { Stats } from '../../hooks/useProfile';
import { HeroProfile } from '../../redux/heroSlice';
import { H3 } from '../../styles/Fonts';
import StatItem from './StatItem';

interface StatsPanelProps {
    stats: Stats;
    setStats: (stats: Stats) => void;
    handleOnSaveProfile: (profile: HeroProfile) => void;
}

const Container = styled.div`
    display: flex;
    justify-content: space-between;

    @media (max-width: 768px) {
        flex-direction: column;
    }
`;

const RightContainer = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    gap: 20px;

    @media (max-width: 768px) {
        margin-top: 40px;
        h3 {
            text-align: right;
        }
    }
`;

function StatsPanel(props: StatsPanelProps) {
    const { stats, setStats, handleOnSaveProfile } = props;

    const handleIncrement = (type: keyof Stats) => {
        if (stats.rest <= 0) return;
        setStats({
            ...stats,
            [type]: stats[type] + 1,
            rest: stats.rest - 1,
        });
    };

    const handleDecrement = (type: keyof Stats) => {
        if (stats[type] <= 0) return;
        setStats({
            ...stats,
            [type]: stats[type] - 1,
            rest: stats.rest + 1,
        });
    };

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { rest, ...statItems } = stats;

    return (
        <Container>
            <div>
                {Object.keys(statItems).map((stat) => (
                    <StatItem
                        key={stat}
                        label={stat.toUpperCase()}
                        point={stats[stat as keyof HeroProfile]}
                        handleIncrement={() =>
                            handleIncrement(stat as keyof HeroProfile)
                        }
                        handleDecrement={() =>
                            handleDecrement(stat as keyof HeroProfile)
                        }
                        restPoint={stats.rest}
                    />
                ))}
            </div>
            <RightContainer>
                <H3>{`剩餘點數: ${stats.rest}`}</H3>
                <Button onClick={() => handleOnSaveProfile(statItems)}>
                    儲存
                </Button>
            </RightContainer>
        </Container>
    );
}

export default StatsPanel;
