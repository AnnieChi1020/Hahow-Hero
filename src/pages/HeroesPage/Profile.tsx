import React, { Fragment } from 'react';
import styled from 'styled-components';
import { useParams } from 'react-router-dom';
import { ClipLoader } from 'react-spinners';
import HeroList from './HeroList';
import StatsPanel from './StatsPanel';
import Colors from '../../styles/Colors';
import useHeroes from '../../hooks/useHeroes';
import useProfile from '../../hooks/useProfile';
import { H3 } from '../../styles/Typography';

const Container = styled.div`
    display: flex;
    flex-direction: column;
    gap: 32px;
    padding: 24px;
`;

const ProfileContainer = styled.div`
    padding: 40px 60px;
    background-color: ${Colors.beige};
    border-radius: 12px;

    @media (max-width: 768px) {
        padding: 40px 40px;
    }
`;

const SubContainer = styled.div`
    width: 100%;
    padding: 60px;
    color: ${Colors.gray5};
    display: flex;
    justify-content: center;
`;

function Profile() {
    const heroList = useHeroes();
    const { stats, setStats, handleOnSaveProfile, isFetching } = useProfile();

    const { heroId } = useParams();

    const showLoader = isFetching && !stats;
    const showProfile = stats;
    const showNoRecord = !isFetching && !stats;

    return (
        <Container>
            <HeroList heroList={heroList} />
            {heroId && (
                <>
                    {showLoader && (
                        <SubContainer>
                            <ClipLoader
                                color={Colors.orange}
                                size={60}
                                data-testid="loader"
                            />
                        </SubContainer>
                    )}
                    {showProfile && (
                        <ProfileContainer>
                            <StatsPanel
                                stats={stats}
                                setStats={setStats}
                                handleOnSaveProfile={handleOnSaveProfile}
                            />
                        </ProfileContainer>
                    )}
                    {showNoRecord && (
                        <SubContainer>
                            <H3>Cannot Find the Profile</H3>
                        </SubContainer>
                    )}
                </>
            )}
        </Container>
    );
}

export default Profile;
