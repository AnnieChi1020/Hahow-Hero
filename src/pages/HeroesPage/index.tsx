import React from 'react';
import styled from 'styled-components';
import HeroList from './HeroList';
import StatsPanel from './StatsPanel';
import Colors from '../../styles/Colors';
import useHeroes from '../../hooks/useHeroes';
import useProfile from '../../hooks/useProfile';

const HeroesPage: React.FC = () => {
    const heroList = useHeroes();
    const { stats, setStats, handleOnSaveProfile } = useProfile();

    return (
        <Container>
            <HeroList heroList={heroList} />
            {stats && (
                <ProfileContainer>
                    <StatsPanel
                        stats={stats}
                        setStats={setStats}
                        handleOnSaveProfile={handleOnSaveProfile}
                    />
                </ProfileContainer>
            )}
        </Container>
    );
};

export default HeroesPage;

const Container = styled.div`
    display: flex;
    flex-direction: column;
    gap: 32px;
    padding: 24px;
`;

const ProfileContainer = styled.div`
    padding: 40px 60px;
    background-color: ${Colors.beige};
    border-radius: 12px;
`;
