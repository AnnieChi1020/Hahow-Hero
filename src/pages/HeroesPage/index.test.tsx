import React from 'react';
import HeroesPage from './index';
import { renderWithProviders } from '../../utils/test-utils';

describe('HeroesPage', () => {
    it('should render correctly', () => {
        const { container } = renderWithProviders(<HeroesPage />);
        expect(container).toMatchSnapshot();
    });
});
