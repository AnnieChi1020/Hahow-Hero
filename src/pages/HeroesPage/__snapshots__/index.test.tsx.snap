// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroesPage should render correctly 1`] = `
<div>
  <div
    class="sc-fmSAUk fOQAxs"
  >
    <div
      class="sc-qRumB dehFKr"
    />
  </div>
</div>
`;
